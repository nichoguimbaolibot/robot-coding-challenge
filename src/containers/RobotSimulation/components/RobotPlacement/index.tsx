import React, { useState } from 'react';

import Button from 'antd/lib/button';
import Input from 'antd/lib/input';
import Select from 'antd/lib/select';

import {
  CARDINAL_DIRECTION,
  ICardinalDirection,
} from 'constant/cardinalDirection';

interface IRobotPlacement {
  onRobotPlacement: (x: number, y: number, f: number) => void;
}

const { Option } = Select;

const RobotPlacement: React.FunctionComponent<IRobotPlacement> = ({
  onRobotPlacement,
}): JSX.Element => {
  const [robotFaceDirection, setRobotFaceDirection] = useState<number>(0);
  const [robotXAxisPlacement, setRobotXAxisPlacement] = useState<number>(0);
  const [robotYAxisPlacement, setRobotYAxisPlacement] = useState<number>(0);

  const handleOnInputXChange = (
    event: React.FormEvent<HTMLInputElement>,
  ): void => {
    const value = event.currentTarget.value;
    setRobotXAxisPlacement(Number(value));
  };

  const handleYAxisPlacement = (
    event: React.FormEvent<HTMLInputElement>,
  ): void => {
    const value = event.currentTarget.value;
    setRobotYAxisPlacement(Number(value));
  };

  const handleRobotDirection = (
    value: React.FormEvent<HTMLSelectElement>,
  ): void => {
    setRobotFaceDirection(Number(value));
  };

  const handlePlaceCommand = (): void => {
    onRobotPlacement(
      robotXAxisPlacement,
      robotYAxisPlacement,
      robotFaceDirection,
    );
  };

  return (
    <div className="robot-placement-section">
      <h4>Place the Robot on table first</h4>

      <Input
        id="input-xAxis"
        placeholder="x-axis position"
        className="input-xAxis"
        onChange={handleOnInputXChange}
      />

      <Input
        id="input-yAxis"
        placeholder="y-axis position"
        className="input-yAxis"
        onChange={handleYAxisPlacement}
      />

      <Select
        id="select-position"
        className="select-position"
        placeholder="ex. North"
        onChange={handleRobotDirection}
      >
        {CARDINAL_DIRECTION.map((cardinal: ICardinalDirection) => {
          return (
            <Option
              className="cardinal-option-value"
              key={cardinal.id}
              value={cardinal.direction}
            >
              {cardinal.position}
            </Option>
          );
        })}
      </Select>

      <Button
        className="place-robot-btn"
        type="primary"
        onClick={handlePlaceCommand}
      >
        PLACE
      </Button>
    </div>
  );
};

export default RobotPlacement;
