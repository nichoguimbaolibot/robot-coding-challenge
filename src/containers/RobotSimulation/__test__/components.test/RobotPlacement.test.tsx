import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import RobotPlacement from 'containers/RobotSimulation/components/RobotPlacement';
import sinon from 'sinon';

afterEach(cleanup);

describe('test ROBOTPLACEMENT component', () => {
  const handleRobotPlacement = sinon.spy();
  it('renders RobotPlacement component', () => {
    const { asFragment } = render(
      <RobotPlacement onRobotPlacement={handleRobotPlacement} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('must have onChangeEvent for x-axis position', () => {
    const testId = 'x-axis-test';
    const { getByTestId } = render(
      <RobotPlacement onRobotPlacement={handleRobotPlacement} />,
    );

    fireEvent.change(getByTestId(testId));
  });

  it('must have onChangeEvent for robot cardinal direction', () => {
    const testId = 'direction-test';
    const { getByTestId } = render(
      <RobotPlacement onRobotPlacement={handleRobotPlacement} />,
    );

    fireEvent.change(getByTestId(testId));
  });

  it('must have onChangeEvent for y-axis position', () => {
    const testId = 'y-axis-test';
    const { getByTestId } = render(
      <RobotPlacement onRobotPlacement={handleRobotPlacement} />,
    );

    fireEvent.change(getByTestId(testId));
  });

  it('must have onClickEvent for ROBOTPLACEMENT position', () => {
    const testId = 'place-robot-test';
    const { getByTestId } = render(
      <RobotPlacement onRobotPlacement={handleRobotPlacement} />,
    );

    fireEvent.click(getByTestId(testId));
  });
});
