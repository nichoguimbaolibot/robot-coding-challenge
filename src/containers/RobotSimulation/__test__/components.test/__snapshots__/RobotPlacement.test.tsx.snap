// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ROBOTPLACEMENT component renders RobotPlacement component 1`] = `
<DocumentFragment>
  <div
    class="robot-placement-section"
  >
    <h4>
      Robot Placement
    </h4>
    <input
      class="ant-input input-xAxis"
      data-testid="x-axis-test"
      id="input-xAxis"
      placeholder="x-axis position"
      type="text"
      value=""
    />
    <input
      class="ant-input input-yAxis"
      data-testid="y-axis-test"
      id="input-yAxis"
      placeholder="y-axis position"
      type="text"
      value=""
    />
    <div
      class="select-position ant-select ant-select-enabled"
      id="select-position"
    >
      <div
        aria-autocomplete="list"
        aria-controls="test-uuid"
        aria-expanded="false"
        aria-haspopup="true"
        class="ant-select-selection
            ant-select-selection--single"
        data-testid="direction-test"
        role="combobox"
        tabindex="0"
      >
        <div
          class="ant-select-selection__rendered"
        >
          <div
            class="ant-select-selection__placeholder"
            style="display: block; user-select: none;"
            unselectable="on"
          >
            ex. North
          </div>
        </div>
        <span
          class="ant-select-arrow"
          style="user-select: none;"
          unselectable="on"
        >
          <i
            aria-label="icon: down"
            class="anticon anticon-down ant-select-arrow-icon"
          >
            <svg
              aria-hidden="true"
              class=""
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </i>
        </span>
      </div>
    </div>
    <button
      class="ant-btn place-robot-btn ant-btn-primary"
      data-testid="place-robot-test"
      type="button"
    >
      <span>
        PLACE
      </span>
    </button>
  </div>
</DocumentFragment>
`;
